pkg/fuzzer/queue/queue.go

```diff
@@ -0,0 +1,375 @@
+import (
+	"context"
+	"math/rand"
+	"sync"
+
+	"github.com/google/syzkaller/pkg/hash"
+	"github.com/google/syzkaller/pkg/ipc"
+	"github.com/google/syzkaller/pkg/signal"
+	"github.com/google/syzkaller/pkg/stats"
+	"github.com/google/syzkaller/prog"
+)
+
+type Request struct {
+	Prog       *prog.Prog
+	NeedSignal SignalType
+	NeedCover  bool
+	NeedHints  bool
+	// TODO: passing ExecOpts directly is actually quite hacky -- this is a high-level interface
+	// and it should specify what is needed in high-level terms.
+	ExecOpts *ipc.ExecOpts
+
+	// If specified, the resulting signal for call SignalFilterCall
+	// will include subset of it even if it's not new.
+	SignalFilter     signal.Signal
+	SignalFilterCall int
+
+	// This stat will be incremented on request completion.
+	Stat *stats.Val
+
+	// Options needed by runtest.
+	BinaryFile   string // If set, it's executed instead of Prog.
+	Repeat       int    // Repeats in addition to the first run.
+	ReturnOutput bool
+
+	// The callback will be called on request completion in the LIFO order.
+	// If it returns false, all further processing will be stopped.
+	// It allows wrappers to intercept Done() requests.
+	callback DoneCallback
+
+	mu          sync.Mutex
+	initialized bool
+	result      *Result
+	done        chan struct{}
+}
+
+type DoneCallback func(*Request, *Result) bool
+
+func (r *Request) OnDone(cb DoneCallback) {
+	oldCallback := r.callback
+	r.callback = func(req *Request, res *Result) bool {
+		r.callback = oldCallback
+		if !cb(req, res) {
+			return false
+		}
+		if oldCallback == nil {
+			return true
+		}
+		return oldCallback(req, res)
+	}
+}
+
+func (r *Request) Done(res *Result) {
+	if r.callback != nil {
+		if !r.callback(r, res) {
+			return
+		}
+	}
+	if r.Stat != nil {
+		r.Stat.Add(1)
+	}
+	r.initChannel()
+	r.result = res
+	close(r.done)
+}
+
+// Wait() blocks until we have the result.
+func (r *Request) Wait(ctx context.Context) *Result {
+	r.initChannel()
+	select {
+	case <-ctx.Done():
+		return &Result{Status: ExecFailure}
+	case <-r.done:
+		return r.result
+	}
+}
+
+func (r *Request) initChannel() {
+	r.mu.Lock()
+	if !r.initialized {
+		r.done = make(chan struct{})
+		r.initialized = true
+	}
+	r.mu.Unlock()
+}
+
+type SignalType int
+
+const (
+	NoSignal  SignalType = iota // we don't need any signal
+	NewSignal                   // we need the newly seen signal
+	AllSignal                   // we need all signal
+)
+
+type Result struct {
+	Info   *ipc.ProgInfo
+	Output []byte
+	Status Status
+	Err    error // More details in case of ExecFailure.
+}
+
+func (r *Result) Stop() bool {
+	return r.Status == ExecFailure || r.Status == Crashed
+}
+
+type Status int
+
+const (
+	Success     Status = 0
+	ExecFailure Status = 1 // For e.g. serialization errors.
+	Crashed     Status = 2 // The VM crashed holding the request.
+	Restarted   Status = 3 // The VM was restarted holding the request.
+)
+
+// Executor describes the interface wanted by the producers of requests.
```

Comment:
Executor arranges call to req.Done after Submit, right? May be a useful detail to mention here.

Reply:
What do you mean exactly by "arranges call to req.Done after Submit"?

Reply:
I meant that whoever calls Submit expects that the executor will call Done on the request sooner later (inside of Submit, or asynchronously later).
It's effectively part of the interface. An Executor implementation that never calls Done on requests is broken.